### PR TITLE
Fix #179 security hardening and #199 default model/CI

### LINKED ISSUE #179: Slack allow list bypass + a few other security fixes
I was reading through the code and noticed a few security items, mostly
  one-line fixes:

  1. Config file permissions — pkg/config/config.go:352 writes the config (which
   contains API keys and bot tokens) with 0644. The auth store already uses 0600
   at pkg/auth/store.go:73, so this looks like an oversight. Same for the cron
  store at pkg/cron/service.go:343.

  2. Slack allow list bypass — handleSlashCommand and handleAppMention in
  pkg/channels/slack.go don't call IsAllowed(), so any user in the workspace can
   use the bot regardless of the allow list config.

  3. Cron exec restriction — pkg/tools/cron.go:36 creates ExecTool with
  restrict=false, meaning scheduled commands aren't confined to the workspace.
  Looks unintentional since the main agent's exec tool respects the restriction
  setting.

  All of these are 1-3 line changes. Happy to send a PR

### PR DESCRIPTION
**Pull request by jmahotiedu**
## Summary
- fix Slack allowlist bypass for app mentions and slash commands (#179)
- harden secret/sensitive file persistence to private permissions (

### CODE REVIEW

**Comment by Leeaandrob** [comment]
@Zepan This PR addresses security hardening (fixing #179) and default model/CI improvements (#199). The roadmap explicitly calls for **Security Audit & Bug Fixes** as volunteer role #3.

**Recommendation: Review and merge.** Security fixes should be prioritized — the roadmap document acknowledges early "vibe coding" created security gaps.

**Comment by Leeaandrob** [comment]
@jmahotiedu take look into the conflicts

File: pkg/channels/slack.go [hunk 1]
```diff
@@ -295,6 +295,12 @@ func (c *SlackChannel) handleAppMention(ev *slackevents.AppMentionEvent) {
 	if ev.User == c.botUserID {
 		return
 	}
+	if !c.IsAllowed(ev.User) {
```
Comment (Leeaandrob) [inline pkg/channels/slack.go:310]: **[Positive — Allowlist Fix for App Mentions]** @jmahotiedu — Correct placement. The `IsAllowed()` check is placed BEFORE the reaction emoji, pending ack, and `HandleMessage()` call. This means blocked users get zero feedback — no "eyes" reaction, no processing. This is the right security posture.

Note: `BaseChannel.HandleMessage` (base.go:86) also has its own `IsAllowed()` check as defense-in-depth, but this early return avoids unnecessary work (reaction API calls, string processing) for blocked users.

File: pkg/channels/slack.go [hunk 2]
```diff
@@ -346,6 +352,14 @@ func (c *SlackChannel) handleSlashCommand(event socketmode.Event) {
 	}
 
 	senderID := cmd.UserID
+	if !c.IsAllowed(senderID) {
+		logger.DebugCF("slack", "Slash command rejected by allowlist", map[string]interface{}{
+			"user_id": senderID,
+			"command": cmd.Command,
```
Comment (Leeaandrob) [inline pkg/channels/slack.go:380]: **[Positive — Allowlist Fix for Slash Commands]** @jmahotiedu — Same correct pattern as app mentions. The check is early, before any processing or response. The `cmd.Command` in the debug log is a good touch for audit trails.

File: pkg/config/config.go [hunk 3]
```diff
@@ -317,3 +319,25 @@ func expandHome(path string) string {
...
 }
+
+func writePrivateFile(path string, data []byte) error {
+	if err := os.WriteFile(path, data, 0600); err != nil {
+		return err
+	}
+	return os.Chmod(path, 0600)
+}
+
+func normalizeLegacyModelDefaults(cfg *Config) {
```
Comment (Leeaandrob) [inline pkg/config/config.go]: **[Positive + M2 Suggestion]** @jmahotiedu — Smart migration function. The logic flow is clear:

1. Not `glm-4.7`? → leave it alone (user chose something deliberately)
2. Has Zhipu key? → leave it (they intend to use Zhipu)
3. Has OpenRouter key? → migrate to `openrouter/auto`
4. Neither key? → **do nothing** (model stays `glm-4.7`)

Case 4 is intentionally conservative (we cannot guess the user's intended provider), but it should have a test to document this behavior:

```go
func TestLoadConfigKeepsGLMWhenNoKeysConfigured(t *testing.T) {
    // ... config with model=glm-4.7, no API keys
    // Assert model is still glm-4.7
}
```

File: pkg/config/config.go [hunk 4]
```diff
@@ -317,3 +319,25 @@ func expandHome(path string) string {
 	}
 	return path
 }
+
+func writePrivateFile(path string, data []byte) error {
```
Comment (Leeaandrob) [inline pkg/config/config.go]: **[M1 — MEDIUM: DRY Violation]** @jmahotiedu — This exact function is duplicated in `pkg/cron/service.go:460`. Both are identical:

```go
func writePrivateFile(path string, data []byte) error {
    if err := os.WriteFile(path, data, 0600); err != nil {
        return err
    }
    return os.Chmod(path, 0600)
}
```

Consider moving it to `pkg/utils/` (which already exists with `media.go` and `string.go`) and importing from both packages. Not blocking, but worth cleaning up.

**Review by Leeaandrob — APPROVED** [review]
## PR #211 Deep Review — @jmahotiedu

Excellent work, @jmahotiedu. This is a well-structured security PR that addresses real vulnerabilities from issue #179 with proper test coverage. The migration strategy for the default model is smart — gracefully handles existing users' configs at load time.

---

### Verification

- Checked out branch locally: `fix/issues-179-199-security-model`
- `go vet ./pkg/channels/... ./pkg/config/... ./pkg/cron/... ./pkg/auth/... ./pkg/migrate/...` — clean
- `go test` on all 5 modified packages — **all 62 tests passing**
- `go test ./pkg/channels/... -race` — clean, no data races
- Confirmed `handleAppMention`, `handleSlashCommand`, and `handleMessageEvent` ALL now have `IsAllowed()` checks
- Confirmed `openrouter/auto` is correctly routed via `strings.HasPrefix(model, "openrouter/")` at http_provider.go:203
- Confirmed `writePrivateFile` correctly handles both new files (via 0600 perm) AND existing files with wrong perms (via explicit `os.Chmod`)
- Audited issue #179 — items 1 (file perms) and 2 (Slack allowlist) are fixed; item 3 (cron exec restriction) appears to have been refactored away in current code architecture

---

### Summary of Findings

**MEDIUM (Suggestions, not blocking)**
- **M1**: `writePrivateFile` is duplicated identically in `config.go:323` and `cron/service.go:460` — DRY violation. Move to `pkg/utils/`.
- **M2**: `normalizeLegacyModelDefaults` — no test for the edge case where `glm-4.7` is set but NEITHER Zhipu nor OpenRouter key exists (model stays `glm-4.7` with no usable provider). Add a test to document this intentional behavior.
- **M3**: CI workflow could benefit from `go vet ./...` step (catches common Go issues beyond formatting) and `-race` flag on `go test`.

**LOW**
- **L1**: Merge conflicts with current upstream/main are expected (Chinese/Japanese READMEs don't exist on this branch's base). Already flagged.

**POSITIVE**
1. **Slack allowlist bypass fixed** — `handleAppMention` (line 298) and `handleSlashCommand` (line 355) now enforce `IsAllowed()` BEFORE any processing (reactions, acks, HandleMessage). Correct placement.
2. **File permissions hardened to 0600** — config.json and cron store both contain API keys/tokens. 0644 was a real leak vector on shared systems.
3. **Smart migration** — `normalizeLegacyModelDefaults` at config load time gracefully migrates existing users without breaking Zhipu users who intentionally use `glm-4.7`.
4. **Test coverage** — every change has corresponding tests: allowlist rejection for app mentions and slash commands, file permission verification, config normalization (both OpenRouter migration AND Zhipu preservation).
5. **Cross-platform test awareness** — `runtime.GOOS == "windows"` skip for permission tests; `setTestHome` helper for HOME env across Linux/Mac/Windows.
6. **CI workflow** — clean gofmt + go test, `go-version-file: go.mod` for version consistency.
7. **`os.Chmod` after `os.WriteFile` is correct** — `WriteFile` doesn't change permissions on existing files, so the chmod is necessary for files that already exist with 0644.

---

### Verdict: APPROVE

Solid security hardening with comprehensive tests. The medium items are suggestions for improvement, not blockers. Ready to merge after resolving upstream merge conflicts.

**Comment by Leeaandrob** [comment]
@jmahotiedu take look on it

**Comment by jmahotiedu** [comment]
Addressed the review suggestions in follow-up commit e2f3b43:

- M1: deduplicated private-file write logic by introducing pkg/utils.WritePrivateFile and using it from both pkg/config/config.go and pkg/cron/service.go.
- M2: added the missing edge-case test TestLoadConfigKeepsLegacyGLMModelWhenNoProviderKeysConfigured in pkg/config/config_test.go.
- M3: updated CI workflow (.github/workflows/ci.yml) to run go vet ./... and go test -race ./....

Local verification run:
- go test ./...
- go test ./pkg/config ./pkg/cron ./pkg/utils
- go vet ./pkg/config ./pkg/cron ./pkg/utils

Note: go test -race could not be executed in my local Windows environment because CGO is disabled by default there; race coverage is now enforced in CI on Linux.

File: pkg/config/config.go (hunk 3 above)
Comment (jmahotiedu) [reply pkg/config/config.go]: Addressed in e2f3b43: added `TestLoadConfigKeepsLegacyGLMModelWhenNoProviderKeysConfigured` in `pkg/config/config_test.go` to document the no-provider-keys glm-4.7 behavior.

File: pkg/config/config.go (hunk 4 above)
Comment (jmahotiedu) [reply pkg/config/config.go]: Addressed in e2f3b43: moved the duplicated helper to `pkg/utils/file.go` as `utils.WritePrivateFile` and switched both config/cron call sites to use it.

**Comment by jmahotiedu** [comment]
Follow-up: merged upstream main into this branch in 11f4a51, resolved the remaining conflicts, and preserved the review-fix changes (utils.WritePrivateFile, legacy-model normalization tests, and CI vet/race checks). The PR is now mergeable again.

**Comment by jmahotiedu** [comment]
Addressed the CI failure in ee798bd by generating embedded workspace assets before vet/tests (go generate ./cmd/picoclaw), so go vet ./... no longer fails on missing workspace embed files in clean CI checkouts.

**Comment by jmahotiedu** [comment]
@Leeaandrob @Zepan

Conflicts are resolved and PR #211 is now mergeable (`mergeable_state: clean`). I pushed the updates to the PR branch (`efb6850`).

What was done:
- Merged latest `main` into `fix/issues-179-199-security-model`
- Resolved conflicts in `README.md`, `config/config.example.json`, `pkg/config/config.go`, `pkg/cron/service.go`
- Kept security hardening behavior (Slack allowlist checks + private-file writes via `utils.WritePrivateFile`)
- Fixed follow-up lint/format issues from the merge

Local verification:
- `go test ./pkg/channels ./pkg/config ./pkg/auth ./pkg/cron ./pkg/migrate` (pass)

Fresh CI is running for this head:
- PR workflow: https://github.com/sipeed/picoclaw/actions/runs/22319073695
- CI workflow: https://github.com/sipeed/picoclaw/actions/runs/22319073787

If these complete green, please merge ASAP.

**Comment by jmahotiedu** [comment]
@Leeaandrob @Zepan

Update: latest head `8da7f08` is now fully green.

- PR workflow: ✅ https://github.com/sipeed/picoclaw/actions/runs/22319254133
- CI workflow: ✅ https://github.com/sipeed/picoclaw/actions/runs/22319256328

PR is conflict-free and mergeable (`mergeable_state: clean`). Please merge when convenient.

**Comment by jmahotiedu** [comment]
﻿@Huaaudio @Leeaandrob quick status update for `#211`:

The WeCom race-test failure is fixed on the current head (`8da7f08`), and the latest checks are all green:
- CI / `test`: https://github.com/sipeed/picoclaw/actions/runs/22319256328
- PR / `Linter` + `Tests`: https://github.com/sipeed/picoclaw/actions/runs/22319254133

GitHub now shows this PR as blocked on review (`REVIEW_REQUIRED`).
Could one of you please do the final review and merge if everything looks good?
